modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java

```diff
@@ -18,135 +18,53 @@
  */
 package org.nuxeo.ecm.core.api.pathsegment;
 
-import java.util.LinkedList;
-
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
 import org.nuxeo.ecm.core.api.DocumentModel;
-import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.runtime.RuntimeServiceException;
 import org.nuxeo.runtime.model.ComponentContext;
-import org.nuxeo.runtime.model.ComponentInstance;
 import org.nuxeo.runtime.model.DefaultComponent;
 
 /**
  * Central service for the generation of a path segment for a document.
  */
 public class PathSegmentComponent extends DefaultComponent implements PathSegmentService {
 
-    private static final Log log = LogFactory.getLog(PathSegmentComponent.class);
-
     public static final String XP = "pathSegmentService";
 
-    protected LinkedList<Class<? extends PathSegmentService>> contribs;
-
     protected PathSegmentService service;
 
-    protected boolean recompute;
-
     @Override
-    public void activate(ComponentContext context) {
-        contribs = new LinkedList<>();
-        recompute = true;
-        service = null;
-    }
-
-    @Override
-    public void deactivate(ComponentContext context) {
-        contribs.clear();
-        service = null;
-    }
-
-    @Override
-    @SuppressWarnings("unchecked")
-    public void registerContribution(Object contrib, String xp, ComponentInstance contributor) {
-        if (!XP.equals(xp)) {
-            log.error("Unknown extension point " + xp);
-            return;
-        }
-        if (!(contrib instanceof PathSegmentServiceDescriptor)) {
-            log.error("Invalid contribution: " + contrib.getClass().getName());
-            return;
-        }
-        PathSegmentServiceDescriptor desc = (PathSegmentServiceDescriptor) contrib;
-        Class<?> klass;
-        try {
-            klass = Class.forName(desc.className);
-        } catch (ClassNotFoundException e) {
-            log.error("Invalid contribution class: " + desc.className);
-            return;
-        }
-        if (!PathSegmentService.class.isAssignableFrom(klass)) {
-            log.error("Invalid contribution class: " + desc.className);
-            return;
+    public void start(ComponentContext context) {
+        PathSegmentService confService = this.<PathSegmentServiceDescriptor> getRegistryContribution(XP).map(desc -> {
+            try {
+                return desc.className.getDeclaredConstructor().newInstance();
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeServiceException(e);
+            }
+        }).orElse(null);
+        if (confService == null) {
+            confService = new PathSegmentServiceDefault();
```

Comment:
Coud be predefined in a constant, and in the `orElse`

Reply:
Sadly using orElse directly does not work here:
```
The method orElse(capture#1-of ? extends PathSegmentService) in the type Optional<capture#1-of ? extends PathSegmentService> is not applicable for the arguments (PathSegmentService)
```

I'll change the assignment still

Reply:
With an additional cast this seems to compile;

        service = this.<PathSegmentServiceDescriptor> getRegistryContribution(XP).map(desc -> {
            try {
                return (PathSegmentService) desc.className.getDeclaredConstructor().newInstance();
            } catch (ReflectiveOperationException e) {
                throw new RuntimeServiceException(e);
            }
        }).orElse(DEFAULT_SERVICE);

And all the lambda could be in a helper `newInstance(PathSegmentServiceDescriptor)` method to make this nicer.

Reply:
I'll keep that in mind when working on https://jira.nuxeo.com/browse/NXP-30092: there are too many places where we do such instantiations (and i'd still like to throw better error messages in cases where the className would be null, which is not the case here).
The related error management should be consistent in all cases (some add runtime error messages, some not, etc...)

Reply:
Just for my knowledge, in such a situation, we prefer to declare the default service as a static member VS doing a `orElseGet(PathSegmentService::new)` for instance?

Reply:
Not necessarily, at least that's not what I understood, it's just syntax-related.
Here i do agree we'd better instantiate at start, it even if it should not make a big difference (as the default class does not hold any state).
It's a bit different when this default value is non-null after the component is stopped (in ExceptionHandler component typically), where using a static member makes more sense to me, or when the instantiated class holds a state (in which case re-instantiating makes more sense).

Reply:
Agreed that here it does not make a big difference.